workflow-api/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+//! This needs to be its own crate so that it doesn't pull in anything that would make compiling
+//! to WASM not work. I've already figured out how to do all that once before with my WASM workflows
+//! hackathon
+
+mod activity_definitions;
+
+use activity_definitions::MyActFnWfCtxExt;
+use futures::future::BoxFuture;
+use std::time::Duration;
+use temporal_sdk_core_protos::{
+    coresdk::workflow_commands::ContinueAsNewWorkflowExecution, temporal::api::common::v1::Payload,
+};
+
+// anyhow errors are used for the errors returned by user-defined functions. This makes `?` work
+// well everywhere by default which is a very nice property, as well as preserving backtraces. We
+// may need to define our own error type instead to allow attaching things like the non-retryable
+// flag... but I suspect we can just make downcasting work for that.
+
+/// Workflow authors must implement this trait to create Temporal Rust workflows
+pub trait Workflow: Sized {
+    /// Type of the input argument to the workflow
+    type Input: TemporalDeserializable;
+    /// Type of the output of the workflow
+    type Output: TemporalSerializable;
+    /// The workflow's name
+    const NAME: &'static str;
+
+    /// Called when an instance of a Workflow is first initialized.
+    ///
+    /// `input` contains the input argument to the workflow as defined by the client who requested
+    /// the Workflow Execution.
+    fn new(input: Self::Input, ctx: SafeWfContext) -> Self;
+
+    /// Defines the actual workflow logic. The function must return a future, and this future is
+    /// cached and polled as updates to the workflow history are received.
+    ///
+    /// `ctx` should be used to perform various Temporal commands like starting timers and
+    /// activities.
+    fn run(
+        &mut self,
+        ctx: WfContext,
+    ) -> BoxFuture<Result<WfExitValue<Self::Output>, anyhow::Error>>;
```

Comment:
Is there any way we can avoid forcing `anyhow` on users here? Can we just have `TemporalFailure` enum here and some into() or whatever-similar-to-anyhow helper that's clear it converts to `ApplicationFailure`?

Reply:
It can certainly be our own thing yes